### PR TITLE
BUG: windows with TemporaryFile an read_csv #13398

### LINKED ISSUE #13398: TemporaryFile as input to read_table raises TypeError: '_TemporaryFileWrapper' object is not an iterator
Although the requirement in the doc says that the input can be a file like object, it doesn't work with objects from tempfile. On Windows, they can't be reopened, so I need to pass the object itself.

#### Code Sample, a copy-pastable example if possible

```
import pandas as pd
from tempfile import TemporaryFile
new_file = TemporaryFile("w+")
dataframe = pd.read_table(new_file, skiprows=3, header=None, sep=r"\s*")
```

#### Expected Output
Not an exception!

#### output of ``pd.show_versions()``
INSTALLED VERSIONS
------------------
commit: None
python: 3.5.1.final.0
python-bits: 64

pandas: 0.18.0

### PR DESCRIPTION
**Pull request by mbrucher**
- [ x] closes #13398 
- [ x] no tests added -> could not find location for IO tests?
- [ x] passes `git diff upstream/master | flake8 --diff`

Change the way of reading back to readline (consistent with the test before entering the function)

One failure on Windows 10 (Python 3.5), but expected to fail actually (should probably tag it as well?)

```
======================================================================
FAIL: test_next (pandas.io.tests.test_common.TestMMapWrapper)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "G:\Informatique\pandas\pandas\io\tests\test_common.py", line 139, in test_next   
    self.assertEqual(next_line, line)
AssertionError: 'a,b,c\r\n' != 'a,b,c\n'
- a,b,c
?      -
+ a,b,c
```

### CODE REVIEW

**Comment by jreback** [comment]
@mbrucher no, pls rebase on master, that test failure was fixed very recently.

I don't like skipping ANYTHING if at all possible (even on windows), yes occasionally we do, but only if its not a simple fix.

**Comment by jreback** [comment]
tests go here: https://github.com/pydata/pandas/blob/master/pandas/io/tests/parser/common.py

these will be testsed on all platforms and parsers. Model after other tests.s @

unless @gfyoung has a better location.

**Comment by mbrucher** [comment]
OK, thx.
Actually, I'm on master from yesterday (missed one patch, I don't know why). The test doesn't seem to be mixed on master :/

**Comment by mbrucher** [comment]
I'll add the test.

**Comment by jreback** [comment]
If you have something failing, pls post; its clean now on appveyor and my manual runs on 3.5 (and 2.7)

**Comment by jreback** [comment]
```
commit 9670b3139171b3b628cecb87a3d0b46cc0361eba
Author: Jeff Reback <jeff@reback.net>
Date:   Tue Jun 14 17:44:13 2016 -0400

    TST: skip mmap error comparison on windows

```

**Comment by gfyoung** [comment]
@jreback : yes, `common.py` is the right place for this to go since it doesn't fall into any nice category that was laid out previously.  BTW, wouldn't consider this a bug but rather a compat issue because you can iterate over a `TemporaryFile`in Python 2.x but not in Python 3.x.

**Comment by mbrucher** [comment]
Oh... That may be because I have the anaconda version of pandas that is used instead of my develop :/
So I'll add the test and we'll see.

**Comment by mbrucher** [comment]
It may actually also fail on OS X and Linux, not only on Windows . Definitely compatibility issue, but also a mismatch in the implementation (testing the availablity of a function and using another one to do the actual work).

**Comment by jreback** [comment]
yeah bug/compat doesn't matter much, except what I think I overuse the compat label :<

either way if ``TemporaryFile`` is not behaving as 'regular' files that's an issue. More generally I would say that that is a python bug no? Or is that defined for some reason?

**Comment by mbrucher** [comment]
I've added a test. I kept read_table so that I could use a regexp with the Python engine and use the path using the generator (sep!=None or len(sep) > 1)

File: pandas/io/tests/test_common.py [hunk 1]
```diff
@@ -90,6 +90,15 @@ def test_iterator(self):
...
 
+    def test_temporary_file(self):
+        from tempfile import TemporaryFile
+        new_file = TemporaryFile("w+")
+        new_file.write("0   0")
+        new_file.flush()
+        new_file.seek(0)
+
+        dataframe = read_table(new_file, sep=r"\s+", header=None, engine="python")
+
```
Comment (gfyoung) [inline pandas/io/tests/test_common.py]: No need to import `read_table` for this.  `read_csv` and `read_table` are absolutely identical (except for the default delimiter).  Also, you should check the result of the `read_` operation to make sure you aren't getting garbage out.  Also make reference the issue (the one you raised).

**Comment by mbrucher** [comment]
Indeed, files work properly. So probably a Python default in the implementation. But the test in pandas was not the good one anyway (should have tested __next__ instead of readline).

File: pandas/io/tests/test_common.py (hunk 1 above)
Comment (mbrucher) [inline pandas/io/tests/test_common.py]: You meqn reference the issue in the test (comment)?

Comment (gfyoung) [inline pandas/io/tests/test_common.py]: Exactly:
~~~python
def test_temporary_file(self):
    # see gh-13398
    ...
~~~

**Comment by gfyoung** [comment]
@jreback : not entirely sure why they changed the behavior of `TemporaryFile` - it's indeed an odd feature but it hasn't really reared its head until now.  With regards to your question about opening and closing again, I'm not sure I follow you here.  When you say re-open with the same name, how?  I don't think you can specify a name with `NamedTemporaryFile`.

File: pandas/io/tests/test_common.py [hunk 2]
```diff
@@ -33,7 +33,8 @@ class TestCommonIOCapabilities(tm.TestCase):
 foo2,12,13,14,15
 bar2,12,13,14,15
 """
-
+    data2 = data1.replace(",", " ")
+    
     def test_expand_user(self):
```
Comment (gfyoung) [inline pandas/io/tests/test_common.py]: I think your original example was fine.  No need to do this here unless `data2` is intended to be used elsewhere.  All you had to do was change `read_table` to `read_csv`.

File: pandas/io/tests/test_common.py (hunk 1 above)
Comment (mbrucher) [inline pandas/io/tests/test_common.py]: Are those tests run? They all fail when called explicitely (and they should).

Comment (gfyoung) [inline pandas/io/tests/test_common.py]: Yes, they do.  If you look at `test_parsers.py`, all of them get imported as test cases.

**Comment by mbrucher** [comment]
@gfyoung No, you can't specify a name indeed. It's if you want to open the file through another mechanism (which is what people use on forums, they don't pass the object, they open the file again, but it fails for Windows if yu do this).

File: pandas/io/tests/test_common.py (hunk 2 above)
Comment (mbrucher) [inline pandas/io/tests/test_common.py]: It's to use the regular expression and make sure I'm going through the code path I modified (I need a non default sep).
Also see my current issue with the file (not run by nosetests by default?).

**Comment by gfyoung** [comment]
@mbrucher : ah, yes, that is indeed true - it fails on Windows but does work on Unix though.

File: pandas/io/tests/test_common.py (hunk 1 above)
Comment (mbrucher) [inline pandas/io/tests/test_common.py]: Some of the errors I get:

nosetests -s pandas\io\tests\parser\python_parser_only.py:PythonParserTests
EEEEEEEE
======================================================================
ERROR: pandas.io.tests.parser.python_parser_only.PythonParserTests.test_BytesIO_input
----------------------------------------------------------------------
Traceback (most recent call last):
  File "G:\Anaconda3\lib\site-packages\nose\case.py", line 198, in runTest
    self.test(*self.arg)
  File "G:\Informatique\pandas\pandas\io\tests\parser\python_parser_only.py", line 84, in test_BytesIO_input
    result = self.read_table(data, sep="::", encoding='cp1255')
AttributeError: 'PythonParserTests' object has no attribute 'read_table'

======================================================================
ERROR: pandas.io.tests.parser.python_parser_only.PythonParserTests.test_decompression_regex_sep
----------------------------------------------------------------------
Traceback (most recent call last):
  File "G:\Anaconda3\lib\site-packages\nose\case.py", line 198, in runTest
    self.test(*self.arg)
  File "G:\Informatique\pandas\pandas\io\tests\parser\python_parser_only.py", line 133, in test_decompression_regex_sep
    data = open(self.csv1, 'rb').read()
AttributeError: 'PythonParserTests' object has no attribute 'csv1'

======================================================================
ERROR: pandas.io.tests.parser.python_parser_only.PythonParserTests.test_negative_skipfooter_raises
----------------------------------------------------------------------
Traceback (most recent call last):
  File "G:\Anaconda3\lib\site-packages\nose\case.py", line 198, in runTest
    self.test(*self.arg)
  File "G:\Informatique\pandas\pandas\io\tests\parser\python_parser_only.py", line 34, in test_negative_skipfooter_raises
    self.read_csv(StringIO(text), skipfooter=-1)
AttributeError: 'PythonParserTests' object has no attribute 'read_csv'

And they are proper issues in the test file.

File: pandas/io/tests/test_common.py (hunk 2 above)
Comment (mbrucher) [inline pandas/io/tests/test_common.py]: OK, figured it out.

Comment (gfyoung) [inline pandas/io/tests/test_common.py]: Your original test was using regular expression.

Comment (mbrucher) [inline pandas/io/tests/test_common.py]: Oh! That's just to build a consistent DataFrame with two different mechanism, I'm assuming the one with StringIO + csv data will always be OK.

File: pandas/io/tests/test_common.py [hunk 3]
```diff
@@ -90,6 +91,18 @@ def test_iterator(self):
...
 
+    def test_temporary_file(self):
+        # GH13398
+        from tempfile import TemporaryFile
+        new_file = TemporaryFile("w+")
+        new_file.write(self.data2)
+        new_file.flush()
+        new_file.seek(0)
+
+        result = read_csv(new_file, sep=r"\s+", engine="python")
```
Comment (jreback) [inline pandas/io/tests/test_common.py]: is there a corresponding test for the c engine that does the same? or is this specifically an issue with the python engine?

Comment (mbrucher) [inline pandas/io/tests/test_common.py]: I think it's specific to the Python engine as it has to do with the regex.

File: pandas/io/tests/parser/python_parser_only.py [hunk 4]
```diff
@@ -171,3 +171,25 @@ def test_read_table_buglet_4x_multiindex(self):
...
+"""
+        data2 = data1.replace(",", " ")
+
+        from tempfile import TemporaryFile
+        new_file = TemporaryFile("w+")
+        new_file.write(data2)
+        new_file.flush()
+        new_file.seek(0)
+
+        result = self.read_csv(new_file, sep=r"\s+", engine="python")
```
Comment (gfyoung) [inline pandas/io/tests/parser/python_parser_only.py]: You don't need to specify the engine here because it's already specified as Python only.  General rule of thumb for these tests: you shouldn't need to explicitly specify the engine.

**Comment by jreback** [comment]
ok, @gfyoung comment, and pls add a whatsnew note. ping when green.

File: pandas/io/tests/parser/python_parser_only.py [hunk 5]
```diff
@@ -171,3 +171,25 @@ def test_read_table_buglet_4x_multiindex(self):
...
+        expected = self.read_csv(StringIO(data1))
```
Comment (gfyoung) [inline pandas/io/tests/parser/python_parser_only.py]: Better practice is to explicitly define your expected output by constructing the DataFrame itself.  If you want, you can simplify your data input to make this construction easier to write out (does your data need to be that long for the error to be triggered - why not that nice simple data you used in your initial commit?)

Comment (mbrucher) [inline pandas/io/tests/parser/python_parser_only.py]: I was trying to be too clever. I reused the data from test_common, should have reverted it back.
Should be good now.

**Comment by jreback** [comment]
thanks @mbrucher ping when green.

File: doc/source/whatsnew/v0.18.2.txt [hunk 6]
```diff
@@ -493,6 +493,7 @@ Bug Fixes
 - Bug in ``pd.read_csv()`` in which the ``nrows`` argument was not properly validated for both engines (:issue:`10476`)
 - Bug in ``pd.read_csv()`` with ``engine='python'`` in which infinities of mixed-case forms were not being interpreted properly (:issue:`13274`)
 - Bug in ``pd.read_csv()`` with ``engine='python'`` in which trailing ``NaN`` values were not being parsed (:issue:`13320`)
+- Bug in ``pd.read_csv()`` with ``engine='python'`` when reading from a tempfile.TemporaryFile on Windows with Python 3 a file with the separator expressed as a regex (:issue:`13398`)
 - Bug in ``pd.read_csv()`` that prevents ``usecols`` kwarg from accepting single-byte unicode strings (:issue:`13219`)
```
Comment (gfyoung) [inline doc/source/whatsnew/v0.18.2.txt]: I think there are some grammar issues in that sentence.

File: pandas/io/tests/parser/python_parser_only.py [hunk 7]
```diff
@@ -171,3 +171,18 @@ def test_read_table_buglet_4x_multiindex(self):
             columns=list('abcABC'), index=list('abc'))
         actual = self.read_table(StringIO(data), sep='\s+')
         tm.assert_frame_equal(actual, expected)
+
+    def test_temporary_file(self):
+        # GH13398
+        data1 = "0,0"
+        data2 = data1.replace(",", " ")
+
```
Comment (gfyoung) [inline pandas/io/tests/parser/python_parser_only.py]: Why do we need to do this?  What's wrong with `data = "0 0"`?

File: doc/source/whatsnew/v0.18.2.txt (hunk 6 above)
Comment (mbrucher) [inline doc/source/whatsnew/v0.18.2.txt]: OK, changed

File: pandas/io/tests/parser/python_parser_only.py (hunk 7 above)
Comment (mbrucher) [inline pandas/io/tests/parser/python_parser_only.py]: Nothing. It's late Sunday evening with a crappy game on tv :p

File: pandas/io/tests/parser/python_parser_only.py [hunk 8]
```diff
@@ -171,3 +171,18 @@ def test_read_table_buglet_4x_multiindex(self):
...
+        from tempfile import TemporaryFile
+        new_file = TemporaryFile("w+")
+        new_file.write(data2)
+        new_file.flush()
+        new_file.seek(0)
+
+        result = self.read_csv(new_file, sep=r"\s+", header=None)
```
Comment (gfyoung) [inline pandas/io/tests/parser/python_parser_only.py]: Can you double check that this actually errors without your changes?  I couldn't get an error on my end, which means you need to change the `sep` argument.

File: pandas/io/tests/parser/python_parser_only.py [hunk 9]
```diff
@@ -171,3 +171,17 @@ def test_read_table_buglet_4x_multiindex(self):
...
+        # GH13398
+        data1 = "0 0"
+
+        from tempfile import TemporaryFile
+        new_file = TemporaryFile("w+")
+        new_file.write(data1)
+        new_file.flush()
+        new_file.seek(0)
+
+        result = self.read_csv(new_file, sep=r"\s+", header=None)
```
Comment (gfyoung) [inline pandas/io/tests/parser/python_parser_only.py]: Can you double check that this actually errors without your changes?  I couldn't get an error on my end (I think you need to change the `sep` argument to another multi-char regex).

Comment (mbrucher) [inline pandas/io/tests/parser/python_parser_only.py]: I have the error when I revert the change.

Comment (gfyoung) [inline pandas/io/tests/parser/python_parser_only.py]: Hmm...how odd.  It's because we do special handling of `\s+`, and you might be seeing it (while I am not) due to LOCALE.  To be safe, I would recommend doing `sep="\s*"`.

File: pandas/io/tests/parser/python_parser_only.py (hunk 7 above)
Comment (gfyoung) [inline pandas/io/tests/parser/python_parser_only.py]: Join the club! :smile:

File: pandas/io/tests/parser/python_parser_only.py (hunk 9 above)
Comment (mbrucher) [inline pandas/io/tests/parser/python_parser_only.py]: OK, I understand. I don't think it makes a difference as the test is not yet done there ont he regex, but I don't know the rest of pandas, so I trust you. I used @jreback regex, so using mine again in my last change.

**Comment by gfyoung** [comment]
Ah, that's better!  @jreback this LGTM and is ready to merge if there are no other concerns, and Travis is happy.  Thanks for your patience @mbrucher !

**Comment by mbrucher** [comment]
Sorry for the earlier rants ;)
I fixed the last lint issues, hopefully the last commits won't break the fix I put in. We'll see tomorrow.

**Comment by mbrucher** [comment]
@jreback Everything is green now, thx for the help!

**Comment by jreback** [comment]
thanks!
